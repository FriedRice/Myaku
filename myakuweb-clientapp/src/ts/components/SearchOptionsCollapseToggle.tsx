/** @module Toggle button component for showing/hiding the search options */

import React from 'react';

interface SearchOptionsCollapseToggleProps {
    collapsed: boolean;
    onToggle: () => void;
}
type Props = SearchOptionsCollapseToggleProps;


const SearchOptionsCollapseToggle: React.FC<Props> = function(props) {
    var buttonText;
    if (props.collapsed) {
        buttonText = 'Show search options';
    } else {
        buttonText = 'Hide search options';
    }

    return (
        <button
            className='button-link search-options-toggle'
            type='button'
            onClick={props.onToggle}
        >
            {buttonText}
        </button>
    );
};

export default SearchOptionsCollapseToggle;
