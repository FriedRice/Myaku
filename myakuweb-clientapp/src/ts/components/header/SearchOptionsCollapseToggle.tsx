/**
 * SearchOptionsCollapseToggle component module. See
 * [[SearchOptionsCollapseToggle]].
 */

import React from 'react';

/** Props for the [[SearchOptionsCollapseToggle]] component. */
interface SearchOptionsCollapseToggleProps {
    /**
     * Whether the search options are currently collapsed or not.
     *
     * @remarks
     * The toggle will indicate that the search options will be shown by
     * toggling if this is true, and the toggle will indicate that the search
     * options will be hidden by toggling if this is false.
     */
    collapsed: boolean;

    /** Callback to call whether the toggle is clicked. */
    onToggle: () => void;
}
type Props = SearchOptionsCollapseToggleProps;

/**
 * Toggle button component for showing/hiding the search options in a form.
 *
 * @param props - See [[SearchOptionsCollapseProps]].
 */
const SearchOptionsCollapseToggle: React.FC<Props> = function(props) {
    var buttonText;
    if (props.collapsed) {
        buttonText = 'Show search options';
    } else {
        buttonText = 'Hide search options';
    }

    return (
        <button
            className='button-link search-options-toggle'
            type='button'
            onClick={props.onToggle}
        >
            {buttonText}
        </button>
    );
};

export default SearchOptionsCollapseToggle;
