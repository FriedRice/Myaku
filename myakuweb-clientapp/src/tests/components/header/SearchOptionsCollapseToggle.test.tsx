/**
 * Tests for the SearchOptionsCollapseToggle component.
 * @module tests/components/header/SearchOptionsCollapseToggle.test
 */

import React from 'react';
import SearchOptionsCollapseToggle from
    'ts/components/header/SearchOptionsCollapseToggle';

import {
    ReactWrapper,
    mount,
} from 'enzyme';

function expectButtonText(wrapper: ReactWrapper, collapsed: boolean): void {
    const button = wrapper.find('button');
    expect(button).toHaveLength(1);

    if (collapsed) {
        expect(button.text()).toStrictEqual(expect.stringContaining('Show'));
        expect(button.text()).not.toStrictEqual(
            expect.stringContaining('Hide')
        );
    } else {
        expect(button.text()).toStrictEqual(expect.stringContaining('Hide'));
        expect(button.text()).not.toStrictEqual(
            expect.stringContaining('Show')
        );
    }
}

function simulateButtonClick(wrapper: ReactWrapper): void {
    const button = wrapper.find('button');
    expect(button).toHaveLength(1);
    button.simulate('click');
}


describe('<SearchOptionsCollapseToggle />', function() {
    it('has text that matches collapsed state', function() {
        const wrapper = mount(
            <SearchOptionsCollapseToggle
                collapsed
                onToggle={(): void => {}}
            />
        );
        expect(wrapper).toMatchSnapshot();
        expectButtonText(wrapper, true);

        wrapper.setProps({'collapsed': false});
        expect(wrapper).toMatchSnapshot();
        expectButtonText(wrapper, false);

        wrapper.setProps({'collapsed': true});
        expect(wrapper).toMatchSnapshot();
        expectButtonText(wrapper, true);
    });

    it('calls onToggle on button click', function() {
        const mockOnToggle = jest.fn();
        const wrapper = mount(
            <SearchOptionsCollapseToggle
                collapsed
                onToggle={mockOnToggle}
            />
        );
        simulateButtonClick(wrapper);
        expect(mockOnToggle).toBeCalledTimes(1);

        simulateButtonClick(wrapper);
        expect(mockOnToggle).toBeCalledTimes(2);

        wrapper.setProps({'collapsed': false});
        simulateButtonClick(wrapper);
        expect(mockOnToggle).toBeCalledTimes(3);

        simulateButtonClick(wrapper);
        expect(mockOnToggle).toBeCalledTimes(4);
    });
});
