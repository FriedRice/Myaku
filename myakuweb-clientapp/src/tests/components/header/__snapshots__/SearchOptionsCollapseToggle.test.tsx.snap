// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchOptionsCollapseToggle /> has text that matches collapsed state 1`] = `
<SearchOptionsCollapseToggle
  collapsed={true}
  onToggle={[Function]}
>
  <button
    className="button-link search-options-toggle"
    onClick={[Function]}
    type="button"
  >
    Show search options
  </button>
</SearchOptionsCollapseToggle>
`;

exports[`<SearchOptionsCollapseToggle /> has text that matches collapsed state 2`] = `
<SearchOptionsCollapseToggle
  collapsed={false}
  onToggle={[Function]}
>
  <button
    className="button-link search-options-toggle"
    onClick={[Function]}
    type="button"
  >
    Hide search options
  </button>
</SearchOptionsCollapseToggle>
`;

exports[`<SearchOptionsCollapseToggle /> has text that matches collapsed state 3`] = `
<SearchOptionsCollapseToggle
  collapsed={true}
  onToggle={[Function]}
>
  <button
    className="button-link search-options-toggle"
    onClick={[Function]}
    type="button"
  >
    Show search options
  </button>
</SearchOptionsCollapseToggle>
`;
